### PR TITLE
chore: Modular configuration system

### PR DESCRIPTION
**Pull request by tdpeuter**
This PR completely overhauls the configuration system, migrating from Tyro to [Hydra](https://hydra.cc/docs/intro/), and introduces foundational support for decentralized control architectures and partial limb amputations.

Using this new configuration method, it should be easier to run different experiments in the (near) future.

- [x] Replaced Tyro with Hydra. Configurations are now backed by strict dataclasses (`BrittleStarConfig`) and hierarchical YAMLs in `configs/`. The way configurations are parsed is now *way* more elegant. (Please trust me.)
- [x] Added support for dynamically swapping between `CentralizedConfig` and `DecentralizedConfig`.
- [x] Amputations: Modified `MorphologyConfig` to accept a list of segments per arm (e.g., `[4, 0, 4, 2, 4]`) to natively support procedural amputations. Also implemented boolean mask scattering (`padded_obs_wrapper.py`) to maintain consistent global state vectors for centralized controllers during amputations.
- [x] Added additional tests and a `--debug_sanity` flag to `train.py`.

## Caveats

- CLI arguments have fundamentally changed. You must now use Hydra override syntax (e.g., `python scripts/train.py ppo.learning_rate=0.005` instead of `--learning-rate 0.005`).
- Legacy `PPOArgs` and flat YAML files have removed.
- Decentralized MLPs (propagator/message passing) still need to be implemented; this PR only sets up the configuration and architecture routing for them.

## Testing

You can verify the configurations, tensor shapes, and MuJoCo procedural generation by running the new tests.

<details>
<summary>Brittle star visualisaties</summary>

amputated_arm
full_5_arm

</details>

@JibrilExe All logs, except the HPC specific logs (with `.e` and `.o` suffix), should be centralized now. You can configure the run-dir with a cli argument.

### CODE REVIEW

**Comment by tdpeuter** [comment]
Hold on, ironing out some issues.

File: src/experiment_logger/unified_logger.py [hunk 1]
```diff
@@ -21,38 +21,67 @@
...
+def init_logger(**kwargs) -> "UnifiedLogger":
+    """Initialize the full UnifiedLogger and set it as the active logger.
+
+    This should be called once the configuration is ready. It will create
+    the output directories and set up all logging backends.
+    """
+    global _active_logger
+    logger = UnifiedLogger(_set_as_global=False, **kwargs)
+    _active_logger = logger
+    return logger
```
[the remark this answers is not in the document]
Comment (tdpeuter) [reply src/experiment_logger/unified_logger.py:56]: Fixed in e40a4b979f859c35fd181addac5156d43e4fdb8b

File: configs/main_config.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+# Sub-configs are loaded from the relative directories.
+
+defaults:
+  - experiment: base
+  - logging: default
+  - ppo: default
+  - architecture: centralized
+  - morphology: 5_arms_full
+  - arena: default
+  - environment: directed_locomotion
```
[the remark this answers is not in the document]
Comment (tdpeuter) [reply configs/main_config.yaml:13]: Fixed in e40a4b979f859c35fd181addac5156d43e4fdb8b

File: tests/test_morphology_render.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    # 1. Compute forward kinematics so geoms are correctly positioned
+    mujoco.mj_forward(model_full, data_full)
+
+    # 2. Render using the environment's primary camera (camera=0)
+    renderer_full = mujoco.Renderer(model=model_full)
+    renderer_full.update_scene(data_full, camera=1)
+    pixels_full = renderer_full.render()
+    image_path = os.path.join(base_dir, "full_5_arm.png")
+    Image.fromarray(pixels_full).save(image_path)
+    print(f"Generated full morphology render: {image_path}")
```
[the remark this answers is not in the document]
Comment (tdpeuter) [reply tests/test_morphology_render.py:39]: Fixed in 5a358fadded334577b1d0395c6dbf62277ef6c34

File: src/brittle_star_project/environment/padded_obs_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        mask_1x.extend([True] * actual + [False] * (ref - actual))
+        # 2x scaling (e.g., joints: 2 values per segment)
+        mask_2x.extend([True] * (actual * 2) + [False] * ((ref - actual) * 2))
+
+    return {
+        "mask_1x": jnp.array(mask_1x, dtype=bool),
+        "mask_2x": jnp.array(mask_2x, dtype=bool),
+        "target_size_1x": sum(reference_segments_per_arm),
+        "target_size_2x": sum(reference_segments_per_arm) * 2,
+    }
```
[the remark this answers is not in the document]
Comment (tdpeuter) [reply src/brittle_star_project/environment/padded_obs_wrapper.py]: Fixed in 143dbbc710ba90496491ad9eaeb045f4ac1f9438

File: src/brittle_star_project/trainers/PPOTrainer.py [hunk 5]
```diff
@@ -500,20 +515,27 @@ def train(self):
...
             )
 
-        if self.args.save_model:
-            model_path = f"{self.run_dir}/{self.args.exp_name}.cleanrl_model"
+            if getattr(self.config.experiment, "debug_sanity", False):
+                self.logger.log(
+                    "\n[SANITY CHECK] Successfully completed 1 epoch of data collection and gradient updates."
+                )
+                self.logger.log("[SANITY CHECK] Gradients flowed without NaN. Exiting gracefully.")
+                break
```
[the remark this answers is not in the document]
Comment (tdpeuter) [reply src/brittle_star_project/trainers/PPOTrainer.py]: Fixed in e40a4b979f859c35fd181addac5156d43e4fdb8b

File: tests/test_configs.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    """Test that the centralized configuration composes and validates correctly."""
+    with initialize(version_base="1.3", config_path="../configs"):
+        # We compose the config; it follows main_config.yaml
+        cfg = compose(config_name="main_config", overrides=["architecture=centralized"])
+
+        # Merge with the dataclass class to get a structured DictConfig,
+        # then convert to a real dataclass instance to verify validation.
+        structured_cfg = OmegaConf.to_object(OmegaConf.merge(BrittleStarConfig, cfg))
+
+        # Basic assertions
```
[the remark this answers is not in the document]
Comment (tdpeuter) [reply tests/test_configs.py]: Fixed in a72471611771af206abc69e8f55f42334119a2bb

File: scripts/train.py [hunk 7]
```diff
@@ -1,75 +1,58 @@
...
-    run_name = f"{args.exp_name}__seed_{args.seed}__{git_hash}__{int(time.time())}"
 
-    if args.run_dir is None:
-        run_dir = f"runs/{run_name}"
-    else:
-        run_dir = args.run_dir
+@hydra.main(config_path="../configs", config_name="main_config", version_base="1.3")
+def main(dict_cfg: DictConfig):
+    # 1. Convert DictConfig to structured dataclass
+    cfg: BrittleStarConfig = OmegaConf.to_object(dict_cfg)
```
[the remark this answers is not in the document]
Comment (tdpeuter) [reply scripts/train.py:29]: Fixed in a72471611771af206abc69e8f55f42334119a2bb

File: scripts/simulate.py [hunk 8]
```diff
@@ -10,86 +18,58 @@
...
+def main(dict_cfg: DictConfig) -> None:
+    cfg: BrittleStarConfig = OmegaConf.to_object(dict_cfg)
 
-    morphology_cfg, arena_cfg, env_cfg = from_file("../configs/test.yaml")
+    # Allow overriding these via Hydra CLI or a dedicated simulate config group
+    # For now, defaults matching the old argparse behavior
+    backend = Backend.MJX
+    model_type = "random"
+    model_path = None
+    seed = cfg.experiment.seed
```
[the remark this answers is not in the document]
Comment (tdpeuter) [reply scripts/simulate.py]: Fixed in a72471611771af206abc69e8f55f42334119a2bb

File: tests/test_configs.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+# Registration must happen before composition to enable validation against schemas
+register_configs()
+
+
+def test_config_composition_centralized():
+    """Test that the centralized configuration composes and validates correctly."""
+    with initialize(version_base="1.3", config_path="../configs"):
+        # We compose the config; it follows main_config.yaml
+        cfg = compose(config_name="main_config", overrides=["architecture=centralized"])
```
[the remark this answers is not in the document]
Comment (tdpeuter) [reply tests/test_configs.py]: Adressed in a7c1aebd09f662f5a1165bf1297958e08a9bfaa0

File: src/brittle_star_project/configs/register_configs.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    cs.store(group="ppo", name="base_ppo", node=PPOConfig)
+
+    # Architecture variants — swap via CLI: architecture=decentralized
+    cs.store(group="architecture", name="centralized", node=CentralizedConfig)
+    cs.store(group="architecture", name="decentralized", node=DecentralizedConfig)
+
+    # Environment configs
+    cs.store(group="morphology", name="base_morphology", node=MorphologyConfig)
+    cs.store(group="arena", name="base_arena", node=ArenaConfig)
+    cs.store(group="environment", name="base_environment", node=EnvConfig)
```
[the remark this answers is not in the document]
Comment (tdpeuter) [reply src/brittle_star_project/configs/register_configs.py:39]: Addressed in a7c1aebd09f662f5a1165bf1297958e08a9bfaa0

**Comment by tdpeuter** [comment]
Verified on the HPC (donphan debug/interactive) with `ppo=fast` as well. [WandB run](https://wandb.ai/SEL3-2026-Groep-4/PPO-Modularity/runs/j061cn2t)

File: src/brittle_star_project/environment/BrittleStarJaxEnvWrapper.py [hunk 11]
```diff
@@ -62,7 +65,13 @@ def reset(self, seed: int = 0):
         self.logger.info(f"Resetting vectorized environment environments with seed {seed}")
         self._action_rng, env_rng = jax.random.split(jax.random.PRNGKey(seed), 2)
         env_rngs = jnp.array(jax.random.split(env_rng, self._num_envs))
-        return self._vectorized_reset(rng=env_rngs)
+        state = self._vectorized_reset(rng=env_rngs)
+        from brittle_star_project.environment.padded_obs_wrapper import pad_observations_batched
```
Comment (RobinMeersman) [inline src/brittle_star_project/environment/BrittleStarJaxEnvWrapper.py]: Maybe import this global as this import is also used in another method.

File: src/experiment_logger/config_utils.py [hunk 12]
```diff
@@ -24,7 +21,7 @@ def load_yaml_config(config_path: str) -> Dict[str, Any]:
     if config is None:
         return {}
 
-    log.info(f"Loaded configuration from: {config_path}")
+    get_logger().info(f"Loaded configuration from: {config_path}")
```
Comment (RobinMeersman) [inline src/experiment_logger/config_utils.py:24]: More a question: what is the advantage of calling the `getlogger` function each time compared to using the variable as before?

**Review by RobinMeersman — CHANGES_REQUESTED** [review]
Looks good! 1 change that needs to be looked at and then good to merge

File: src/experiment_logger/config_utils.py (hunk 12 above)
Comment (tdpeuter) [reply src/experiment_logger/config_utils.py:24]: By invoking *get()* every time, a proxy is used. The proxy properly handles when no logging specs/config have been provided yet. It basically allows changing the underlying log*ger* on the fly.

Comment (RobinMeersman) [reply src/experiment_logger/config_utils.py:24]: Noted!

**Review by RobinMeersman — APPROVED** [review]
